src/strategies/erc721-with-metadata-attribute/index.ts

```diff
@@ -0,0 +1,97 @@
+import { Multicaller } from '../../utils';
+import { BigNumber } from '@ethersproject/bignumber';
+import snapshots from '@snapshot-labs/snapshot.js';
+
+export const author = 'allmysmarts';
+export const version = '0.1.0';
+
+const abi = [
+  'function balanceOf(address account) external view returns (uint256)',
+  'function tokenOfOwnerByIndex(address owner, uint256 index) external view returns (uint256)',
+  'function tokenURI(uint256 tokenId) public view returns (string uri)'
+];
+
+export async function strategy(
+  space,
+  network,
+  provider,
+  addresses,
+  options,
+  snapshot
+) {
+  const blockTag = typeof snapshot === 'number' ? snapshot : 'latest';
+
+  // 1st, get the balance of the token
+  const callWalletToBalanceOf = new Multicaller(network, provider, abi, {
+    blockTag
+  });
+  for (const walletAddress of addresses) {
+    callWalletToBalanceOf.call(walletAddress, options.address, 'balanceOf', [
+      walletAddress
+    ]);
+  }
+  const walletToBalanceOf: Record<
+    string,
+    BigNumber
+  > = await callWalletToBalanceOf.execute();
+
+  // 2nd, get tokenIds for each address, and index
+  const callWalletIdToTokenID = new Multicaller(network, provider, abi, {
+    blockTag
+  });
+  for (const [walletAddress, count] of Object.entries(walletToBalanceOf)) {
+    if (count.toNumber() > 0) {
+      for (let index = 0; index < count.toNumber(); index++) {
+        callWalletIdToTokenID.call(
+          walletAddress.toString() + '-' + index.toString(),
+          options.address,
+          'tokenOfOwnerByIndex',
+          [walletAddress, index]
+        );
+      }
+    }
+  }
+  const walletIdToTokenID: Record<
+    string,
+    BigNumber
+  > = await callWalletIdToTokenID.execute();
+
+  // 3rd, get the tokenURI for each token
+  const callWalletIdToTokenURI = new Multicaller(network, provider, abi, {
+    blockTag
+  });
+  for (const [walletId, tokenId] of Object.entries(walletIdToTokenID)) {   
+    callWalletIdToTokenURI.call(
+      walletId.toString(),
+      options.address,
+      'tokenURI',
+      [tokenId]
+    );
+  }
+  const walletIdToTokenURI: Record<
+    string, // address + index
+    string  // tokenURI
+  > = await callWalletIdToTokenURI.execute();
+
+  // 4th, get the attributes, and parse the specified field from tokenURI
+  const walletToAttributeValue = {} as Record<string, number>;
+  for (const [walletId, tokenURI] of Object.entries(walletIdToTokenURI)) {
+    const data = await snapshots.utils.getJSON(tokenURI);
```

Comment:
I just noticed that this can add a lot of loads to the score API  server if there are so many tokens in inventory and when all the owners vote, :(

Is there a way to limit the number of requests here, the actual limit is 5 requests, 
https://github.com/snapshot-labs/snapshot-strategies#checklist-for-adding-a-new-strategy

how many tokens did you get in your contract? maybe is there a way to get all the metadata in a single IPFS file?? I think that will be the best way,
maybe pass `metadata` from the `options` which gets the metadata for us

Reply:
In the beginning, I only used 1st token only, that you requested updates for that.
I prefer using the first 2 tokens only for voting.

Reply:
In my contract, it is ok to use 1st token only in case of multiple holdings.

Reply:
This way won't work, if there are 1000 voters on a proposal, this can send 1000+ requests

Reply:
for our case, it is specific, and metadata can be changed after minted...
so we should get metadata every time.

I can not find any difference with other strategies...
for example, in erc-20-balance-of, it is also getting the balance in every votes?

Reply:
For those strategies they call multicall and they can get all data in single request even if 500 addresses votes 

I see that there is `tokenURI` method in your contract right? Which returns single json of all metadata? Maybe you can use it?

Reply:
tokenURI returns single json for each token, not combined one.

Reply:
based on ERC721

Reply:
and all other strategies, using `fetch` should get data for every addresses, not at once...

Reply:
we only use one nft per address, so getting 1st nft's metadata is ok for us.

Reply:
I mean like this: https://6242dbddb6734894c157cfc0.mockapi.io/api/nft/
@allmysmarts

Reply:
> we only use one NFT per address, so getting 1st NFT's metadata is ok for us.

Getting one metadata per request is not ok for snapshot, there is a limit on number of requests sent from a strategy, 
https://github.com/snapshot-labs/snapshot-strategies#checklist-for-adding-a-new-strategy

Reply:
ok. got it.
I will recreate a new PR.
Thank you for your information.
@ChaituVR